rpc_util.go

```diff
@@ -289,53 +289,63 @@ func (p *parser) recvMsg(maxReceiveMessageSize int) (pf payloadFormat, msg []byt
 
 // encode serializes msg and prepends the message header. If msg is nil, it
 // generates the message header of 0 message length.
-func encode(c Codec, msg interface{}, cp Compressor, cbuf *bytes.Buffer, outPayload *stats.OutPayload) ([]byte, error) {
+func encode(c Codec, msg interface{}, cp Compressor, outPayload *stats.OutPayload) ([]byte, error) {
 	var (
-		b      []byte
-		length uint
+		b          []byte
+		length     uint
+		payloadLen = 1
+		sizeLen    = 4
 	)
-	if msg != nil {
-		var err error
-		// TODO(zhaoq): optimize to reduce memory alloc and copying.
-		b, err = c.Marshal(msg)
-		if err != nil {
-			return nil, Errorf(codes.Internal, "grpc: error while marshaling: %v", err.Error())
-		}
-		if outPayload != nil {
-			outPayload.Payload = msg
-			// TODO truncate large payload.
-			outPayload.Data = b
-			outPayload.Length = len(b)
-		}
-		if cp != nil {
-			if err := cp.Do(cbuf, b); err != nil {
-				return nil, Errorf(codes.Internal, "grpc: error while compressing: %v", err.Error())
-			}
-			b = cbuf.Bytes()
+	if msg == nil {
+		var buf = make([]byte, payloadLen+sizeLen)
+		return buf, nil
+	}
+
+	var err error
+	// TODO(zhaoq): optimize to reduce memory alloc and copying.
+	b, err = c.Marshal(msg)
+	if err != nil {
+		return nil, Errorf(codes.Internal, "grpc: error while marshaling: %v", err.Error())
+	}
+	if outPayload != nil {
+		outPayload.Payload = msg
+		// TODO truncate large payload.
+		outPayload.Data = b
+		outPayload.Length = len(b)
+	}
+
+	var cbuf *bytes.Buffer
+
+	if cp != nil {
+		// pre-alloc place
+		cbuf = bytes.NewBuffer(make([]byte, payloadLen+sizeLen))
+		if err := cp.Do(cbuf, b); err != nil {
+			return nil, Errorf(codes.Internal, "grpc: error while compressing: %v", err.Error())
 		}
+		b = cbuf.Bytes()
 		length = uint(len(b))
```

Comment:
you should subtract `payloadLen+sizeLen` here, and then `} else { length = uint(len(b)) }`